From c0c50b253de5027f201ec34384f5ecb9a3f56e1d Mon Sep 17 00:00:00 2001
From: Sveinung Kvilhaugsvik <sveinung84@users.sourceforge.net>
Date: Thu, 11 May 2017 15:23:59 +0200
Subject: [PATCH] PACKET_WEB_CITY_INFO_ADDITION: be more careful.

Give the packet web_city_info_addition a field with the id of the city the
additional info is for. Cancel it where PACKET_CITY_INFO is canceled.

See hrm Feature #659446
---
 client/packhand.c             | 4 ++--
 common/networking/packets.def | 6 ++++--
 server/citytools.c            | 2 ++
 3 files changed, 8 insertions(+), 4 deletions(-)

diff --git a/client/packhand.c b/client/packhand.c
index 6e1324d..d005e3e 100644
--- a/client/packhand.c
+++ b/client/packhand.c
@@ -895,8 +895,8 @@ void handle_city_info(const struct packet_city_info *packet)
   for it.
   TODO: Do not generate code calling this in C-client.
 ****************************************************************************/
-void handle_web_city_info_addition(int granary_size, int granary_turns,
-                                   int buy_gold_cost)
+void handle_web_city_info_addition(int id, int granary_size,
+                                   int granary_turns, int buy_gold_cost)
 {
 }
 
diff --git a/common/networking/packets.def b/common/networking/packets.def
index bdf431f..135254b 100644
--- a/common/networking/packets.def
+++ b/common/networking/packets.def
@@ -637,7 +637,7 @@ end
 
 /************** City packets **********************/
 
-PACKET_CITY_REMOVE = 30; sc, dsend, lsend, cancel(PACKET_CITY_INFO), cancel(PACKET_CITY_SHORT_INFO)
+PACKET_CITY_REMOVE = 30; sc, dsend, lsend, cancel(PACKET_CITY_INFO), cancel(PACKET_WEB_CITY_INFO_ADDITION), cancel(PACKET_CITY_SHORT_INFO)
   CITY city_id;
 end
 
@@ -704,7 +704,7 @@ PACKET_CITY_INFO = 31; sc, lsend, is-game-info, force, cancel(PACKET_CITY_SHORT_
   ESTRING name[MAX_LEN_CITYNAME];
 end
 
-PACKET_CITY_SHORT_INFO = 32; sc, lsend, is-game-info, cancel(PACKET_CITY_INFO)
+PACKET_CITY_SHORT_INFO = 32; sc, lsend, is-game-info, cancel(PACKET_CITY_INFO), cancel(PACKET_WEB_CITY_INFO_ADDITION)
   CITY id; key
   TILE tile;
 
@@ -2223,6 +2223,8 @@ end
 /* Use range 256:511 for these                             */
 
 PACKET_WEB_CITY_INFO_ADDITION = 256; sc, lsend, is-game-info, force, cancel(PACKET_CITY_SHORT_INFO)
+  CITY id; key
+
   UINT16 granary_size;
   TURN granary_turns;
   UINT16 buy_gold_cost;
diff --git a/server/citytools.c b/server/citytools.c
index e72a7ba..06ece51 100644
--- a/server/citytools.c
+++ b/server/citytools.c
@@ -2492,6 +2492,8 @@ void package_city(struct city *pcity, struct packet_city_info *packet,
   } improvement_iterate_end;
 
 #ifdef FREECIV_WEB
+  web_packet->id = pcity->id;
+
   web_packet->granary_size = city_granary_size(city_size_get(pcity));
   web_packet->granary_turns = city_turns_to_grow(pcity);
   web_packet->buy_gold_cost = city_production_buy_gold_cost(pcity);
-- 
2.1.4

